File: frontend/src/types/question.ts
export type Category =
  | 'Verbal Ability'
  | 'Numerical Reasoning'
  | 'Analytical Ability'
  | 'Clerical Operations'
  | 'General Information';

export type Difficulty = 'easy' | 'medium' | 'hard';

export interface Option {
  id: string; // "A", "B", "C", "D"
  text: string;
}

export interface Concept {
  title: string;
  rule_explanation: string;
  key_points?: string[];
  heuristics?: string[];
}

export interface WorkedExample {
  problem: string;
  solution_steps: string[];
  pattern_recognition_note?: string;
}

export interface LearningMetadata {
  concept: Concept;
  worked_example: WorkedExample;
  guided_hint: string;
  misconception_notes?: string;
  reinforcement_question_ids?: string[];
}

export interface Question {
  id: string;
  text: string;
  options: Option[];
  correct_answer: string; // "A", "B", etc.
  category: Category;
  subcategory?: string;
  difficulty: Difficulty;
  hint?: string;
  explanation: string;
  wrong_answer_explanations?: Record<string, string>; // Maps "A", "B", etc. to text
  tags: string[];
  source?: string;
  version?: number;
  learning_metadata?: LearningMetadata;
  metadata?: Record<string, any>;
}

export interface ExamSession {
  id: string;
  type: 'Mock' | 'Practice' | 'Learn';
  status: 'active' | 'completed' | 'abandoned';
  startTime: string;
  endTime?: string;
  questions: Question[];
  userAnswers: Record<string, string>; // questionId -> optionId ("A", "B", etc.)
  score?: number;
}

File: backend/app/models/question.py
from marshmallow import Schema, fields, validate

class OptionSchema(Schema):
    id = fields.Str(required=True)
    text = fields.Str(required=True)

class ConceptSchema(Schema):
    title = fields.Str(required=True)
    rule_explanation = fields.Str(required=True)
    key_points = fields.List(fields.Str())
    heuristics = fields.List(fields.Str())

class WorkedExampleSchema(Schema):
    problem = fields.Str(required=True)
    solution_steps = fields.List(fields.Str(), required=True)
    pattern_recognition_note = fields.Str()

class LearningMetadataSchema(Schema):
    concept = fields.Nested(ConceptSchema, required=True)
    worked_example = fields.Nested(WorkedExampleSchema, required=True)
    guided_hint = fields.Str(required=True)
    misconception_notes = fields.Str()
    reinforcement_question_ids = fields.List(fields.Str()) # References to other question IDs

class QuestionSchema(Schema):
    id = fields.Str(dump_only=True)
    text = fields.Str(required=True)
    options = fields.List(fields.Nested(OptionSchema), required=True, validate=validate.Length(min=2))
    correct_answer = fields.Str(required=True)
    category = fields.Str(required=True, validate=validate.OneOf([
        'Verbal Ability', 'Numerical Reasoning', 'Analytical Ability',
        'Clerical Operations', 'General Information'
    ]))
    subcategory = fields.Str()
    difficulty = fields.Str(required=True, validate=validate.OneOf(['easy', 'medium', 'hard']))
    hint = fields.Str()
    explanation = fields.Str(required=True)
    wrong_answer_explanations = fields.Dict(keys=fields.Str(), values=fields.Str())
    tags = fields.List(fields.Str())
    learning_metadata = fields.Nested(LearningMetadataSchema, allow_none=True)
    created_at = fields.DateTime(dump_only=True)
    updated_at = fields.DateTime(dump_only=True)
